File: apps/androidtv/core/designsystem/src/main/kotlin/dev/spidola/tv/core/designsystem/ActionableErrorView.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

@file:Suppress("MatchingDeclarationName") // Named for the public composable; the data type is a helper.

package dev.spidola.tv.core.designsystem

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.tv.material3.ClickableSurfaceDefaults
import androidx.tv.material3.MaterialTheme
import androidx.tv.material3.Surface
import androidx.tv.material3.Text

/** One labelled action button for an actionable error. */
data class SpidolaErrorButton(
    val title: String,
    val onClick: () -> Unit,
)

/**
 * Presents a failure with its plain-language class, a one-sentence message, and a **non-empty** set
 * of actions (PRD §6.3). "No action available" is unrepresentable: [primary] is a single required
 * button, so the view always offers at least one thing to do — an error dead-end can never be
 * rendered. The primary button takes initial focus.
 */
@Composable
fun ActionableErrorView(
    failureClass: String,
    message: String,
    primary: SpidolaErrorButton,
    modifier: Modifier = Modifier,
    others: List<SpidolaErrorButton> = emptyList(),
) {
    val primaryFocus = remember { FocusRequester() }
    LaunchedEffect(Unit) { primaryFocus.requestFocus() }
    Column(
        modifier = modifier.fillMaxSize().padding(SpidolaSpacing.xl),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(SpidolaSpacing.l, Alignment.CenterVertically),
    ) {
        Text(
            text = failureClass,
            style = MaterialTheme.typography.titleLarge,
            color = SpidolaPalette.BroadcastWhite,
            textAlign = TextAlign.Center,
        )
        Text(
            text = message,
            style = MaterialTheme.typography.bodyLarge,
            color = SpidolaPalette.Static,
            textAlign = TextAlign.Center,
            modifier = Modifier.widthIn(max = 900.dp),
        )
        Row(horizontalArrangement = Arrangement.spacedBy(SpidolaSpacing.m)) {
            ErrorButton(primary, isPrimary = true, modifier = Modifier.focusRequester(primaryFocus))
            others.forEach { button -> ErrorButton(button, isPrimary = false) }
        }
    }
}

@Composable
private fun ErrorButton(
    button: SpidolaErrorButton,
    isPrimary: Boolean,
    modifier: Modifier = Modifier,
) {
    Surface(
        onClick = button.onClick,
        modifier = modifier,
        shape = ClickableSurfaceDefaults.shape(shape = SpidolaFocus.cardShape),
        colors =
            ClickableSurfaceDefaults.colors(
                containerColor = if (isPrimary) SpidolaPalette.TestCardAmber else SpidolaPalette.Set,
                contentColor = if (isPrimary) SpidolaPalette.Studio else SpidolaPalette.BroadcastWhite,
            ),
        scale = SpidolaFocus.scale(),
        border = SpidolaFocus.border(),
    ) {
        Text(
            text = button.title,
            style = MaterialTheme.typography.bodyLarge,
            modifier = Modifier.padding(horizontal = SpidolaSpacing.l, vertical = SpidolaSpacing.m),
        )
    }
}

File: apps/tvos/Packages/DesignSystem/Sources/DesignSystem/ActionableErrorView.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import SwiftUI

/// One labelled action button for an actionable error.
public struct SpidolaErrorButton: Identifiable, Sendable {
  public let id = UUID()
  public let title: String
  public let action: @MainActor () -> Void

  public init(title: String, action: @escaping @MainActor () -> Void) {
    self.title = title
    self.action = action
  }
}

/// Presents a failure with its plain-language class, a one-sentence message, and a **non-empty**
/// set of actions (PRD §6.3). "No action available" is unrepresentable: `primary` is a single
/// required button, so the view always offers at least one thing to do — an error dead-end can
/// never be rendered.
public struct ActionableErrorView: View {
  private let failureClass: String
  private let message: String
  private let primary: SpidolaErrorButton
  private let others: [SpidolaErrorButton]

  @FocusState private var focused: UUID?

  public init(
    failureClass: String,
    message: String,
    primary: SpidolaErrorButton,
    others: [SpidolaErrorButton] = []
  ) {
    self.failureClass = failureClass
    self.message = message
    self.primary = primary
    self.others = others
  }

  private var buttons: [SpidolaErrorButton] { [primary] + others }

  public var body: some View {
    VStack(spacing: SpidolaSpacing.l) {
      Image(systemName: "exclamationmark.triangle.fill")
        .font(.system(size: 56))
        .foregroundStyle(SpidolaPalette.streamRed)
        .accessibilityHidden(true)
      Text(failureClass)
        .font(SpidolaType.title)
        .foregroundStyle(SpidolaPalette.broadcastWhite)
        .multilineTextAlignment(.center)
      Text(message)
        .font(SpidolaType.body)
        .foregroundStyle(SpidolaPalette.staticGray)
        .multilineTextAlignment(.center)
        .frame(maxWidth: 900)
      HStack(spacing: SpidolaSpacing.m) {
        ForEach(buttons) { button in
          Button(button.title, action: button.action)
            .buttonStyle(.plain)
            .padding(.horizontal, SpidolaSpacing.l)
            .padding(.vertical, SpidolaSpacing.m)
            .background(
              button.id == primary.id ? SpidolaPalette.testCardAmber : SpidolaPalette.set
            )
            .foregroundStyle(
              button.id == primary.id ? SpidolaPalette.studio : SpidolaPalette.broadcastWhite
            )
            .font(SpidolaType.body)
            .focused($focused, equals: button.id)
            .spidolaFocusRing(isFocused: focused == button.id)
        }
      }
      .padding(.top, SpidolaSpacing.s)
    }
    .padding(SpidolaSpacing.xl)
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(SpidolaPalette.studio)
    .onAppear { focused = primary.id }
    .accessibilityElement(children: .contain)
    // The words are the caller's; only the punctuation joining them is this layer's, so the format
    // goes through the catalog and a language that wants the halves the other way round can say so.
    .accessibilityLabel(String(localized: "\(failureClass). \(message)", bundle: .module))
  }
}
